### PR TITLE
k051316 tile flip configuration bits, replaces flag setting in callbacks

### PR DESCRIPTION
**Pull request by furrtek**
Use of previously undocumented configuration bits instead of hardcoded per-game tile flags.

Maybe it would be better to have `m_flip*_enabled` variables instead of reading m_ctrlram[0x0e] each time get_tile_info() is called ?

### CODE REVIEW

File: src/devices/video/k051316.cpp [hunk 1]
```diff
@@ -228,6 +229,9 @@ u8 k051316_device::rom_r(offs_t offset)
 
 void k051316_device::ctrl_w(offs_t offset, u8 data)
 {
+	if ((offset == 0x0e) && (data != m_ctrlram[0x0e]))
```
Comment (angelosa) [inline src/devices/video/k051316.cpp:232]: `m_ctrlram` is indeed legacy code. And without a mask on bits 1-2 this will have performance issue if a given client pings the other bits, making these three (including the ROM readback flag) explicit booleans sounds like a good idea to me.

**Review by cuavas — COMMENTED** [review]
Just as a sanity check, have you checked that all drivers using this device are updated as necessary?  Driver sources using the device that aren’t changed by this PR:
* konami/ajax.cpp
* konami/divebomb.cpp
* konami/ultraman.cpp
* konami/wecleman.cpp
* vsystem/tail2nos.cpp

Also, does the `K051316_CB_MEMBER` still need the `flags` parameter?  If no per-driver glue for flags is necessary, it can be removed.

I agree with @angelosa that you should mask the register 0x0E values for comparison to avoid unnecessary tilemap invalidation.

**Comment by cuavas** [comment]
> Just as a sanity check, have you checked that all drivers using this device are updated as necessary?

After a quick check of those source files, it seems none of them set tile flip flags.  Also, a quick check suggests that the `flags` parameter can in fact be eliminated.

File: src/devices/video/k051316.cpp [hunk 2]
```diff
@@ -228,7 +235,17 @@ u8 k051316_device::rom_r(offs_t offset)
...
 void k051316_device::ctrl_w(offs_t offset, u8 data)
 {
+	if (offset == 0x0e)
+	{
+		m_readout_enabled = ~BIT(data, 0);
+		m_flipx_enabled = BIT(data, 1);
+		m_flipy_enabled = BIT(data, 2);
+		if ((data & 0x06) != (m_ctrlram[0x0e] & 0x06))
+			m_tmap->mark_all_dirty();
+	}
```
Comment (cuavas) [inline src/devices/video/k051316.cpp:245]: The trouble with adding these extra members is you now have duplicated state.  That’s a path to bugs.

Also, this doesn’t actually work.  You need `BIT(~data, 0)` or `!BIT(data, 0)` – `~BIT(data, 0)` will always be `true` when converted to `bool`.

File: src/devices/video/k051316.cpp [hunk 3]
```diff
@@ -186,7 +196,7 @@ void k051316_device::device_start()
 
 void k051316_device::device_reset()
 {
-	memset(m_ctrlram, 0, 0x10);
+	std::fill(std::begin(m_ctrlram), std::end(m_ctrlram), 0);
 }
```
Comment (cuavas) [inline src/devices/video/k051316.cpp:200]: I just realised this causes a change in behaviour.  Previously, readout enable and flip bits were part of this array, so they would be cleared on device reset.  With them no longer in the array, they’re unaffected by reset.  Should they be set to initial values on reset, or only on start?

Comment (galibert) [reply src/devices/video/k051316.cpp:200]: The chip has no reset line, for meh :-)

Comment (furrtek) [reply src/devices/video/k051316.cpp:200]: Indeed, strangely enough. Because of that all games seem to configure all registers at startup / after a reset.
